lib/backend.js

```diff
@@ -125,6 +137,8 @@ Backend.prototype.connect = function(connection, req) {
  * (The agent is available through all middleware)
  */
 Backend.prototype.listen = function(stream, req) {
+  // TODO: When a Connection reconnects with a persisted src, it will change the
+  // Agent's ID to be something different than what it had in the connection middleware
```

Comment:
Reminder that we should discuss this. I can't think of a particularly good way of avoiding it (we could potentially set `Agent` to initially have a `null` ID? At least that would be less surprising than it changing?).

We could potentially only trigger the middleware on the `hs` event, but obviously that would break for older clients, so probably not a good idea. We could add an _extra_ middleware, and deprecate this one? But again can't think of a good way of shimming without triggering middlewares twice.

Reply:
Options we discussed:
- Keep connect as-is, null out the ID. Breaking change, if people are looking at the id.
- Add a "client init" or "handshake" middleware. Problem is, the agent id changing after connect could still be considered a breaking change, if someone has stored the id somewhere.
- Delay emitting the connect event until we know there's a confirmed id. Problem - this is tricky to do for older clients that don't go through the handshake. We could potentially do something like emit "connect" just prior to processing the first message and/or after some timeout, but that gets messy.

Feels like the best way longer term is to do some form of breaking change, but I'm not sure if there's some fully backwards-compatible shim we can do in the meantime, as you say. Let's sleep on this for a day or two.

Reply:
This week we discussed:

- add an `agent.src` field, which is initialised to `null`, and then given the connection ID provided by the new handshake
- `agent.src` will remain `null` for legacy clients (who never send an updated ID)
- `agent.clientId` is still set on construction, and never changes
- any internal ShareDB code that references `agent.clientId` should be updated to use some sort of getter method, which will provide `agent.src || agent.clientId`

This way, we can keep our change non-breaking, but still allow the client to suggest an ID. Anyone who cares can look out for the new `src` field as the "canonical" ID.